internal/kargo/promotion_builder.go

```diff
@@ -0,0 +1,259 @@
+		// the Promotion.
+		step.As = generatePromotionTaskStepAlias(taskAlias, step.GetAlias(i))
+
+		// With the variables validated and mapped, they are now available to
+		// the Config of the step during the Promotion execution.
+		step.Vars = vars
+
+		// Append the inflated step to the list of steps.
+		steps = append(steps, *step)
+	}
+	return steps, nil
+}
+
+// getTaskSpec retrieves the PromotionTaskSpec for the given PromotionTaskReference.
+func (b *PromotionBuilder) getTaskSpec(
+	ctx context.Context,
+	project string,
+	ref *kargoapi.PromotionTaskReference,
+) (*kargoapi.PromotionTaskSpec, error) {
+	var spec kargoapi.PromotionTaskSpec
+
+	if ref == nil {
+		return nil, errors.New("missing task reference")
+	}
+
+	switch ref.Kind {
+	case "PromotionTask", "":
+		task := &kargoapi.PromotionTask{}
+		if err := b.client.Get(ctx, client.ObjectKey{Namespace: project, Name: ref.Name}, task); err != nil {
+			return nil, err
+		}
+		spec = task.Spec
+	case "ClusterPromotionTask":
+		task := &kargoapi.ClusterPromotionTask{}
+		if err := b.client.Get(ctx, client.ObjectKey{Name: ref.Name}, task); err != nil {
+			return nil, err
+		}
+		spec = task.Spec
+	default:
+		return nil, fmt.Errorf("unknown task reference kind %q", ref.Kind)
+	}
+
+	return &spec, nil
+}
+
+// generatePromotionName generates a name for the Promotion by combining the
+// Stage name, a ULID, and a short hash of the Freight.
+//
+// The name has the format of:
+//
+//	<stage-name>.<ulid>.<short-hash>
+func generatePromotionName(stageName, freight string) string {
+	if stageName == "" || freight == "" {
+		return ""
+	}
+
+	shortHash := freight
+	if len(shortHash) > shortHashLength {
+		shortHash = shortHash[0:shortHashLength]
+	}
+
+	shortStageName := stageName
+	if len(stageName) > maxStageNamePrefixLength {
+		shortStageName = shortStageName[0:maxStageNamePrefixLength]
+	}
+
+	parts := []string{shortStageName, ulid.Make().String(), shortHash}
+	return strings.ToLower(strings.Join(parts, nameSeparator))
+}
+
+// generatePromotionTaskStepAlias generates an alias for a PromotionTask step
+// by combining the task alias and the step alias.
+func generatePromotionTaskStepAlias(taskAlias, stepAlias string) string {
+	return fmt.Sprintf("%s%s%s", taskAlias, PromotionAliasSeparator, stepAlias)
+}
+
+// promotionTaskVarsToStepVars validates the presence of the PromotionTask
+// variables and maps them to variables which can be used by the inflated
+// PromotionStep.
+func promotionTaskVarsToStepVars(
+	taskVars, promoVars, stepVars []kargoapi.PromotionVariable,
+) ([]kargoapi.PromotionVariable, error) {
+	if len(taskVars) == 0 {
+		return nil, nil
+	}
+
+	promoVarsMap := make(map[string]kargoapi.PromotionVariable, len(promoVars))
+	for _, v := range promoVars {
+		promoVarsMap[v.Name] = v
+	}
+
+	stepVarsMap := make(map[string]kargoapi.PromotionVariable, len(stepVars))
+	for _, v := range stepVars {
+		stepVarsMap[v.Name] = v
+	}
```

Comment:
Is there any edge case here where we lose anything important through the use of maps?

I'm thinking something like this, which seems dumb _but possible_, might be such a case?

```
vars:
- name: foo
  value: bar
- name: foo
  value: ${{ vars.foo + vars.foo }}
```

Reply:
The second would clobber the first in the map, but the result would be an error, I think, instead of "barbar".

Reply:
This is still an issue, the plan is to inflate them as:

1. task-level vars
2. task-step-level vars

Which would allow the task-step-level to overwrite (or utilize) the task-level variable.

The only concern I have with this approach is that the promotion-level variables then can not overwrite the task-level (default) variables, because they are evaluated before any of the variables in the above list (and a task-level variable with the same name would thus overwrite this value).